server/ctrlsubsonic/ctrl.go

```diff
@@ -223,43 +224,47 @@ func withRequiredParams(next http.Handler) http.Handler {
 	})
 }
 
-func withUser(dbc *db.DB) handlerutil.Middleware {
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			params := r.Context().Value(CtxParams).(params.Params)
-			// ignoring errors here, a middleware has already ensured they exist
-			username, _ := params.Get("u")
-			password, _ := params.Get("p")
-			token, _ := params.Get("t")
-			salt, _ := params.Get("s")
-
-			passwordAuth := token == "" && salt == ""
-			tokenAuth := password == ""
-			if tokenAuth == passwordAuth {
-				_ = writeResp(w, r, spec.NewError(10,
-					"please provide `t` and `s`, or just `p`"))
-				return
-			}
-			user := dbc.GetUserByName(username)
-			if user == nil {
-				_ = writeResp(w, r, spec.NewError(40,
-					"invalid username %q", username))
+func (c *Controller) withUser(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		params := r.Context().Value(CtxParams).(params.Params)
+		// ignoring errors here, a middleware has already ensured they exist
+		username, _ := params.Get("u")
+		password, _ := params.Get("p")
+		token, _ := params.Get("t")
+		salt, _ := params.Get("s")
+
+		passwordAuth := token == "" && salt == ""
+		tokenAuth := password == ""
+		if tokenAuth == passwordAuth {
+			_ = writeResp(w, r, spec.NewError(10,
+				"please provide `t` and `s`, or just `p`"))
+			return
+		}
+		user := c.dbc.GetUserByName(username)
+
+		var credsOk bool
+		if tokenAuth && user != nil {
+			credsOk = checkCredsToken(user.Password, token, salt)
+		} else if user != nil {
+			credsOk = checkCredsBasic(user.Password, password)
+		}
+		if !credsOk {
```

Comment:
not sure i like the idea of "if user/pass fails, try LDAP" 

seems like it would be better to have

```
if ldap is enabled {
   do ldap
   return
}
do normal user name password stuff
```

this would require that LDAP is configured from the very start (see other comment)

Reply:
So, just double checking, you want only LDAP or internal authentication at any given point, right?  And if LDAP authentication is enabled, should the default admin user still exist -- or should there be a separate admin filter (which would allow for more than one admin user) that is used to validate users?

Reply:
hmm i'm not sure honestly. i do like the idea of multiple admin roles, and an admin filter. though gonic will always create an admin user when it runs the first db migrations. so that's probably hard to avoid

Reply:
Fixed in 6ca68b2c260d855b8965b762621efd9396c16107, and I've added an admin filter in 5638ef5df7346adf28335c0c125100004416e876.